### PR TITLE
Implemented the save reservation feature and overlap check

### PR DESCRIPTION
**Pull request by Faka-GLB**
Implemented the save reservation functionality along with the overlap check of existing reservations.
Also implemented unit tests of new and modified methods

image

kotlin project tree

### CODE REVIEW

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt [hunk 1]
```diff
@@ -12,11 +17,45 @@ class ReservePresenter (private val view: ReserveActivityContract.ReserveView):
...
+    }
+
+    override fun setReservationEndDate(endDate: Calendar) {
+        model.setEndDate(endDate)
+        view.setEndDateTextView(getDateString(model.getEndDate()))
+    }
+
+    private fun getDateString(date: Calendar?): String {
+        var ret = ConstantUtils.EMPTY_STRING
+        if (date != null) {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt]: change this line by date?.let {

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/DateTimePickerView.kt [hunk 2]
```diff
@@ -2,13 +2,35 @@ package com.onboarding.parkingsystemkotlin.mvp.view
 
 import androidx.fragment.app.DialogFragment
 import com.onboarding.parkingsystemkotlin.databinding.DateTimePickerReservationFragmentBinding
+import com.onboarding.parkingsystemkotlin.listener.DateTimeListener
 import com.onboarding.parkingsystemkotlin.mvp.contract.DateTimePickerContract
 import com.onboarding.parkingsystemkotlin.mvp.view.base.FragmentView
+import java.util.*
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/DateTimePickerView.kt]: do not use wildcard imports pls

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/ReserveView.kt [hunk 3]
```diff
@@ -23,7 +23,49 @@ class ReserveView(activity: Activity, binding: ActivityNewReservationBinding) :
...
+        binding.tvNewReservationActivityStartDate.text = date
+    }
+
+    override fun setEndDateTextView(date: String) {
+        binding.tvNewReservationActivityEndDate.text = date
+    }
+
+    override fun getParkingLot(): Int {
+        val lot: String = binding.editTextNewReservationActivityLotNumber.text.toString()
+        if (lot.isEmpty()) {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/ReserveView.kt]: you can improve this making 

```
return if (lot.isEmpty()) {
           ConstantUtils.PARKING_LOT_NOT_SET
        } else {
           lot.toInt()
}
```

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt (hunk 1 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt]: done

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/DateTimePickerView.kt (hunk 2 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/DateTimePickerView.kt]: every time I used a Calendar on a new file, Kotlin imported java.util.*  that one must've slipped trough

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/ReserveView.kt (hunk 3 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/ReserveView.kt]: done

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/DateTimePickerView.kt (hunk 2 above)
Comment (SartiMau) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/view/DateTimePickerView.kt]: you can configure the imports in the AS preferences, lets check it later, please remind me

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt [hunk 4]
```diff
@@ -12,11 +17,45 @@ class ReservePresenter (private val view: ReserveActivityContract.ReserveView):
...
+
+    override fun setReservationEndDate(endDate: Calendar) {
+        model.setEndDate(endDate)
+        view.setEndDateTextView(getDateString(model.getEndDate()))
+    }
+
+    private fun getDateString(date: Calendar?): String {
+        var ret = ConstantUtils.EMPTY_STRING
+        date?.let {
+            ret = (date.get(Calendar.DAY_OF_MONTH).toString() + ConstantUtils.SLASH +
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt]: here, inside the let, you should use it instead of date

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt (hunk 4 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt]: done

**Review by SartiMau — APPROVED** [review]

File: app/src/main/java/com/onboarding/parkingsystemkotlin/activity/ReserveActivity.kt [hunk 5]
```diff
@@ -4,20 +4,28 @@ import android.content.Context
...
 import androidx.appcompat.app.AppCompatActivity
+import com.onboarding.parkingsystemkotlin.database.ParkingDatabase
 import com.onboarding.parkingsystemkotlin.databinding.ActivityNewReservationBinding
+import com.onboarding.parkingsystemkotlin.listener.DateTimeListener
 import com.onboarding.parkingsystemkotlin.mvp.contract.ReserveActivityContract
+import com.onboarding.parkingsystemkotlin.mvp.model.ReserveModel
 import com.onboarding.parkingsystemkotlin.mvp.presenter.ReservePresenter
 import com.onboarding.parkingsystemkotlin.mvp.view.ReserveView
+import com.onboarding.parkingsystemkotlin.utils.ReservationVerifier
+import java.util.*
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/parkingsystemkotlin/activity/ReserveActivity.kt]: pls remove wild import

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt [hunk 6]
```diff
@@ -12,11 +17,45 @@ class ReservePresenter (private val view: ReserveActivityContract.ReserveView):
...
+
+    override fun setReservationEndDate(endDate: Calendar) {
+        model.setEndDate(endDate)
+        view.setEndDateTextView(getDateString(model.getEndDate()))
+    }
+
+    private fun getDateString(date: Calendar?): String {
+        var ret = ConstantUtils.EMPTY_STRING
+        date?.let {
+            ret = (it.get(Calendar.DAY_OF_MONTH).toString() + ConstantUtils.SLASH +
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt]: In kotlin to concat you can do this, for example you have this: 
val a: String = "Hola"
val b: String = "Mundo"
val c: String = "$a $b"
and avoid using val c: String = a + " " + b

File: app/src/main/java/com/onboarding/parkingsystemkotlin/activity/ReserveActivity.kt (hunk 5 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/activity/ReserveActivity.kt]: done

File: app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt (hunk 6 above)
Comment (Faka-GLB) [reply app/src/main/java/com/onboarding/parkingsystemkotlin/mvp/presenter/ReservePresenter.kt]: done

**Review by BQuinteros — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Pull request merged**